### PR TITLE
Skip the serializer for null and scalar values on normalize

### PR DESCRIPTION
**Pull request by teohhanhui**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no (but see note below)
| Deprecations? | no
| Tickets       | fixes https://github.com/api-platform/core/issues/2593#issuecomment-476881738, fixes https://github.com/api-platform/core/issues/2031
| License       | MIT
| Doc PR        | N/A

Notable changes:

- The serializer is no longer called for `null` and scalar values during normalization. This is consistent with the behaviour in Symfony Serializer's `AbstractObjectNormalizer`: https://github.com/symfony/symfony/blob/v4.3.6/src/Symfony/Component/Serializer/Normalizer/AbstractObjectNormalizer.php#L194-L196
- No more "child inherited" nonsense.

Potential BC breaks:

- If anyone has been using custom normalizers for `null` and scalar values, those custom normalizers will no longer be called. I think that is really an edge case, while the performance win of this change would be felt by the majority of our users.

TODO:

- [x] Rebase onto `master`
- [ ] Blackfire profile

### CODE REVIEW

File: tests/Serializer/AbstractItemNormalizerTest.php [hunk 1]
```diff
@@ -835,8 +835,6 @@ public function testChildInheritedProperty()
 
         $serializerProphecy = $this->prophesize(SerializerInterface::class);
         $serializerProphecy->willImplement(NormalizerInterface::class);
-        $serializerProphecy->normalize('foo', null, Argument::type('array'))->willReturn('foo')->shouldBeCalled();
-        $serializerProphecy->normalize(null, null, Argument::type('array'))->willReturn(null)->shouldBeCalled();
```
Comment (soyuka) [inline tests/Serializer/AbstractItemNormalizerTest.php]: but now this may break some stuff no? :p

Comment (teohhanhui) [reply tests/Serializer/AbstractItemNormalizerTest.php]: See OP:

> Potential BC breaks:
> 
> - If anyone has been using custom normalizers for null and scalar values, those custom normalizers will no longer be called. I think that is really an edge case, while the performance win of this change would be felt by the majority of our users.

**Comment by dunglas** [comment]
Could you provide a Blackfire profile to see if it is worth it?

**Comment by dunglas** [comment]
Also, it must target master, not 2.5. BC breaks shouldn't be accepted even in master... but they definitely can't be shipped with patch releases.

**Comment by teohhanhui** [comment]
I'd say it's a bug that we were even calling the serializer for normalizing `null` and scalar values, but sure, I'll target `master` if that's what you want...

**Comment by Taluu** [comment]
I may have had just a quick look, but what if you want to remove a relationship (e.g a "nullable relation", or "nullable value" actually ?

(Disclaimer, again : Just had a quick look)

**Comment by teohhanhui** [comment]
@Taluu:
> what if you want to remove a relationship?

Not relevant, as this change only affects normalization, not denormalization.

**Comment by Taluu** [comment]
Oh, then what if you want to expose a `null` to say "there is no relation" here ? At least when I'm consuming an api, I'm pretty used to have this kind of "value" rather than having to check if the key is set. 

I may be wrong though.

**Comment by teohhanhui** [comment]
> Oh, then what if you want to expose a null to say "there is no relation" here ? At least when I'm consuming an api, I'm pretty used to have this kind of "value" rather than having to check if the key is set.

It'll remain the same as before. This PR does not change any of that. The only potential BC break is if someone had custom normalizers for null / scalar values. Very highly unlikely.

**Comment by Taluu** [comment]
Had one for denormalization so I guess that yes, shouldn't be a bother. 😄

File: src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml [hunk 2]
```diff
@@ -49,11 +49,6 @@
             <argument type="service" id="api_platform.property_info" />
         </service>
 
-        <service id="api_platform.metadata.property.name_collection_factory.inherited" class="ApiPlatform\Core\Metadata\Property\Factory\InheritedPropertyNameCollectionFactory" decorates="api_platform.metadata.property.name_collection_factory" decoration-priority="10" public="false">
-            <argument type="service" id="api_platform.metadata.resource.name_collection_factory" />
-            <argument type="service" id="api_platform.metadata.property.name_collection_factory.inherited.inner" />
-        </service>
```
Comment (Taluu) [inline src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml]: shouldn't these services be marked as deprecated instead ?

Comment (teohhanhui) [reply src/Bridge/Symfony/Bundle/Resources/config/metadata/metadata.xml]: This service was only decorating another service. It was never intended to be used directly.

**Comment by soyuka** [comment]
Why did you close?

**Comment by teohhanhui** [comment]
Continued in #3248 (targeting `2.5`) using a safer approach. Will open another PR targeting `master` after this, for the removal of "child inherited" metadata.

We will not skip the serializer for `null` and scalar property values.
